File: sdk/ruby/lib/oresoftware/next_loggers/shutdown.rb
# frozen_string_literal: true

module ORESoftware
  module NextLoggers
    # Raised when one or more transports fail during a flush or close.
    #
    # It carries every failure rather than the first, because shutdown is
    # exactly when a single broken destination must not be allowed to hide the
    # state of the others.
    class ShutdownError < RuntimeError
      attr_reader :errors

      def initialize(operation, errors)
        @errors = errors.freeze
        details = errors.map { |error| "#{error.class}: #{error.message}" }.join("; ")
        super("#{operation} failed on #{errors.length} transport(s): #{details}")
      end
    end

    DEFAULT_SHUTDOWN_TIMEOUT = 5.0

    module_function

    # Monotonic deadline, immune to wall-clock steps during a slow shutdown.
    def shutdown_deadline(timeout)
      return nil if timeout.nil?

      Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout.to_f
    end

    def shutdown_remaining(deadline)
      return nil if deadline.nil?

      [deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC), 0.0].max
    end

    # Invokes one lifecycle hook on one transport, passing the remaining budget
    # only to transports that accept it. A transport with a fixed arity keeps
    # working unchanged; one that opts in gets to honour the deadline.
    def call_transport_hook(transport, hook, remaining, *args)
      return :absent unless transport.respond_to?(hook)

      method = transport.method(hook)
      accepts_timeout =
        method.parameters.any? { |kind, name| name == :timeout && %i[key keyreq].include?(kind) }

      if accepts_timeout
        method.call(*args, timeout: remaining)
      else
        method.call(*args)
      end
    end

    # Chains SIGTERM/SIGINT and at_exit to a logger's close.
    #
    # Nothing is installed by requiring this file: a host that embeds the SDK
    # owns its own lifecycle. The previous trap is chained after the drain
    # rather than replaced, and a second signal arriving mid-drain restores the
    # default disposition and re-raises, so a wedged transport can never make
    # the process unkillable.
    class ProcessHooks
      attr_reader :hooks, :reason

      def initialize(hooks, reason, uninstall)
        @hooks = hooks.freeze
        @reason = reason
        @uninstall = uninstall
      end

      def call
        @uninstall.call
      end
      alias uninstall call
    end

    def install_process_hooks!(logger, timeout: DEFAULT_SHUTDOWN_TIMEOUT, force_on_second_signal: true)
      # Deliberately plain flags rather than a Mutex: Ruby raises ThreadError
      # for Mutex#synchronize inside a trap context, and under the GVL a boolean
      # assignment is already indivisible with respect to a trap handler.
      state = { draining: false, active: true }
      previous_traps = {}

      drain = lambda do
        begin
          logger.close(timeout: timeout)
        rescue StandardError => error
          warn("next-loggers shutdown drain failed: #{error.message}")
        end
      end

      # at_exit blocks cannot be unregistered in Ruby, so uninstall flips a flag
      # the block reads. The block outlives uninstall; its effect does not.
      at_exit do
        drain.call if state[:active]
      end

      hooks = ["at-exit"]
      reason = ""

      %w[TERM INT].each do |name|
        signal_hook = name == "TERM" ? "signal-sigterm" : "signal-sigint"
        begin
          previous_traps[name] = Signal.trap(name) do
            if state[:draining] && force_on_second_signal
              # The operator is done waiting. Trapping the signal overrode the
              # kernel default, so returning here would leave the process
              # unkillable for as long as a wedged transport takes to give up.
              Signal.trap(name, "DEFAULT")
              Process.kill(name, Process.pid)
              next
            end
            state[:draining] = true

            # A trap handler may not block on a Mutex, and the drain does. A
            # thread started from trap context may, so the work happens there.
            Thread.new do
              drain.call
              state[:active] = false
              previous = previous_traps[name]
              if previous.respond_to?(:call)
                previous.call
              else
                # Nothing else claimed the signal. Trapping it suppressed the
                # kernel default, so the process would otherwise survive a
                # SIGTERM it was supposed to die from.
                Signal.trap(name, "DEFAULT")
                Process.kill(name, Process.pid)
              end
            end
          end
          hooks << signal_hook
        rescue ArgumentError, Errno::EINVAL => error
          reason = "#{name} trap rejected by the runtime: #{error.message}"
        end
      end

      uninstall = lambda do
        state[:active] = false
        previous_traps.each do |name, previous|
          begin
            Signal.trap(name, previous.nil? ? "DEFAULT" : previous)
          rescue ArgumentError, Errno::EINVAL
            # The platform withdrew the signal; nothing left to restore.
          end
        end
        previous_traps.clear
      end

      ProcessHooks.new(hooks, reason, uninstall)
    end
  end
end

File: sdk/python/src/next_loggers/shutdown.py
"""Bounded drains and opt-in process lifecycle hooks for next-loggers."""

from __future__ import annotations

import atexit
import inspect
import os
import signal
import sys
import threading
import time
from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple

DEFAULT_TIMEOUT = 5.0


class ShutdownError(RuntimeError):
    """Every transport failure seen during one drain, so none is lost to the first."""

    def __init__(self, operation: str, errors: Sequence[BaseException]) -> None:
        super().__init__(
            "{0} failed on {1} transport(s): {2}".format(
                operation,
                len(errors),
                "; ".join(
                    "{0}: {1}".format(type(error).__name__, error) for error in errors
                ),
            )
        )
        self.operation = operation
        self.errors = list(errors)


def deadline_from(timeout: Optional[float]) -> Optional[float]:
    """Turn a relative budget into a monotonic instant; ``None`` means unbounded."""
    if timeout is None:
        return None
    return time.monotonic() + max(0.0, float(timeout))


def remaining_from(deadline: Optional[float]) -> Optional[float]:
    if deadline is None:
        return None
    return deadline - time.monotonic()


def _transport_name(transport: Any) -> str:
    return str(getattr(transport, "name", type(transport).__name__))


def _accepts_timeout(hook: Callable[..., Any]) -> bool:
    try:
        parameters = inspect.signature(hook).parameters
    except (TypeError, ValueError):
        return False
    for parameter in parameters.values():
        if parameter.kind is inspect.Parameter.VAR_KEYWORD:
            return True
        if parameter.name == "timeout" and parameter.kind in (
            inspect.Parameter.POSITIONAL_OR_KEYWORD,
            inspect.Parameter.KEYWORD_ONLY,
        ):
            return True
    return False


def call_transport_hook(
    transport: Any,
    hook_name: str,
    deadline: Optional[float],
    args: Sequence[Any] = (),
) -> Optional[BaseException]:
    """Run one optional transport hook inside whatever budget is left.

    The budget is shared across transports rather than granted per call, so a
    slow first transport shortens — and never extends — the drain that follows.
    """
    hook = getattr(transport, hook_name, None)
    if not callable(hook):
        return None
    remaining = remaining_from(deadline)
    if remaining is not None and remaining <= 0:
        return TimeoutError(
            "shutdown budget expired before {0}.{1}()".format(
                _transport_name(transport), hook_name
            )
        )
    try:
        if remaining is not None and _accepts_timeout(hook):
            hook(*args, timeout=remaining)
        else:
            hook(*args)
    except Exception as error:
        return error
    return None


def raise_if_failed(operation: str, errors: Sequence[BaseException]) -> None:
    if errors:
        raise ShutdownError(operation, errors)


def _restore_and_reraise(signum: int) -> None:
    """Hand the signal back to the kernel default so the host stays killable."""
    signal.signal(signum, signal.SIG_DFL)
    os.kill(os.getpid(), signum)


class ProcessHooks:
    """Disposer returned by :func:`install_process_hooks`; call it to uninstall.

    ``hooks`` names what was actually attached and ``reason`` explains any hook
    the runtime refused, because a caller off the main thread silently gets an
    atexit-only drain that would otherwise look identical to the full install.
    """

    def __init__(
        self,
        hooks: Sequence[str],
        reason: str,
        uninstall: Callable[[], None],
    ) -> None:
        self.hooks: Tuple[str, ...] = tuple(hooks)
        self.reason = reason
        self._uninstall = uninstall

    def __call__(self) -> None:
        self._uninstall()


def install_process_hooks(
    logger: Any,
    *,
    timeout: float = DEFAULT_TIMEOUT,
    force_on_second_signal: bool = True,
) -> ProcessHooks:
    """Drain ``logger`` on interpreter exit, SIGTERM, and SIGINT.

    Nothing is installed by importing this module: hosts that embed the SDK own
    their own lifecycle. Previous SIGTERM/SIGINT handlers are chained after the
    drain instead of being replaced, and a second signal arriving while the
    drain is still running restores the default disposition and re-raises, so a
    wedged transport can never make the process unkillable.
    """
    state_lock = threading.Lock()
    draining = False
    previous_handlers: Dict[int, Any] = {}

    def drain() -> None:
        try:
            logger.close(timeout=timeout)
        except Exception as error:
            print(
                "next-loggers shutdown drain failed: {0}".format(error),
                file=sys.stderr,
            )

    def on_exit() -> None:
        drain()

    def on_signal(signum: int, frame: Any) -> None:
        nonlocal draining
        with state_lock:
            repeated = draining
            draining = True
        if repeated and force_on_second_signal:
            _restore_and_reraise(signum)
            return
        drain()
        previous = previous_handlers.get(signum)
        if callable(previous) and previous not in (signal.SIG_DFL, signal.SIG_IGN):
            previous(signum, frame)

    def uninstall() -> None:
        atexit.unregister(on_exit)
        for signum, previous in list(previous_handlers.items()):
            if signal.getsignal(signum) is on_signal:
                signal.signal(signum, previous)
        previous_handlers.clear()

    atexit.register(on_exit)
    hooks: List[str] = ["atexit"]
    reason = ""

    # signal.signal() is only legal on the main thread of the main interpreter.
    if threading.current_thread() is threading.main_thread():
        for attribute, hook in (("SIGTERM", "signal-sigterm"), ("SIGINT", "signal-sigint")):
            signum = getattr(signal, attribute, None)
            if signum is None:
                reason = "{0} is not available on this platform".format(attribute)
                continue
            try:
                previous_handlers[signum] = signal.signal(signum, on_signal)
            except (OSError, RuntimeError, ValueError) as error:
                reason = "{0} handler rejected by the runtime: {1}".format(attribute, error)
            else:
                hooks.append(hook)
    else:
        reason = (
            "signal handlers were skipped: install_process_hooks ran off the main thread"
        )

    return ProcessHooks(hooks, reason, uninstall)


__all__ = [
    "DEFAULT_TIMEOUT",
    "ProcessHooks",
    "ShutdownError",
    "call_transport_hook",
    "deadline_from",
    "install_process_hooks",
    "raise_if_failed",
    "remaining_from",
]
